common/metrics/go-metrics-m3.go

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package metrics
+
+import (
+	"fmt"
+	"time"
+
+	gometrics "github.com/rcrowley/go-metrics"
+	"github.com/uber-common/bark"
+)
+
+const (
+	delayToFirstLogDump  = time.Minute * 15
+	delayBetweenLogDumps = time.Minute * 60
+	delayBetweenExports  = time.Second * 10
+)
+
+// goMetricsExporter allows an rcrowley/go-metrics registry to be
+// periodically exported to the Cherami metrics system
+type goMetricsExporter struct {
+	c     Client
+	scope int
+
+	// metricNameToMetric maps a metric name to the metric symbol,
+	// e.g. "request-rate" -> OutputhostCGKafkaRequestRate
+	metricNameToMetric  map[string]int
+	closeCh             <-chan struct{}
+	metricNameToLastVal map[string]int64
+	l                   bark.Logger
+	registry            gometrics.Registry // Registry to be exported
+	lastLogDump         time.Time
+}
+
+// NewGoMetricsExporter starts the exporter loop to export go-metrics to the Cherami
+// metrics system and returns the go-metrics registry that should be used
+func NewGoMetricsExporter(
+	c Client,
+	scope int,
+	metricNameToMetric map[string]int,
+	l bark.Logger,
+	closeCh <-chan struct{},
+) gometrics.Registry {
+	// Make a new registy and give it to the caller. This allows the exporters to operate independently
+	registry := gometrics.NewRegistry()
+	r := &goMetricsExporter{
+		c:                   c,
+		scope:               scope,
+		registry:            registry,
+		metricNameToMetric:  metricNameToMetric,
+		l:                   l,
+		closeCh:             closeCh,
+		metricNameToLastVal: make(map[string]int64),
+
+		// Calculate the time to provide the delayToFirstLogDump in export()
+		lastLogDump: time.Now().Add(-1 * delayBetweenLogDumps).Add(delayToFirstLogDump),
+	}
+	go r.run()
```

Comment:
I'm trying to keep this quite simple on the calling side, i.e. just an init and it handles itself, i.e. magic. I can see the argument in following an established pattern, but that seems like it just adds a few more lines of code on the caller side, and it's speculative (i.e. it presumes that the caller would want to stop the metric reporting at some time other than the natural end of its life). Let's talk in-person.